tests/dragonfly/replication_test.py

```diff
@@ -2339,6 +2339,31 @@ async def test_announce_ip_port(df_factory):
     assert port == "1337"
 
 
+@pytest.mark.asyncio
+async def test_replication_timeout_on_full_sync(df_factory: DflyInstanceFactory):
+    # setting replication_timeout to a very small value to force the replica to timeout
+    master = df_factory.create(replication_timeout=100, vmodule="replica=2,dflycmd=2")
+    replica = df_factory.create()
+
+    df_factory.start_all([master, replica])
+
+    c_master = master.client()
+    c_replica = replica.client()
+
+    # runing dubeg populate with lots of itmes ensures that replication timeout will occure on full sync stage
+    async def debug_populate_master():
+        await c_master.execute_command("debug", "populate", "1000000", "foo", "500")
+
+    populate_task = asyncio.create_task(debug_populate_master())
+    await asyncio.sleep(0.2)  # wait for debug popultae start running
+    await c_replica.execute_command(f"REPLICAOF localhost {master.port}")
+    await populate_task
+    await asyncio.sleep(1)
+
+    await assert_replica_reconnections(replica, 0)
```

Comment:
I am slightly confused with this. How can it ever be 0 if we `unconditionally increment it in MainReplFiber` 

```
  while (state_mask_.load() & R_ENABLED) {
    // Discard all previous errors and set default error handler.
    cntx_.Reset([this](const GenericError& ge) { this->DefaultErrorHandler(ge); });
    // 1. Connect socket.
    if ((state_mask_.load() & R_TCP_CONNECTED) == 0) {
      ThisFiber::SleepFor(500ms);
      if (is_paused_)
        continue;
      ec = ResolveHostDns();
      if (ec) {
        LOG(ERROR) << "Error resolving dns to " << server().host << " " << ec;
        continue;
      }

      // Give a lower timeout for connect, because we're
      reconnect_count_++;
      ec = ConnectAndAuth(absl::GetFlag(FLAGS_master_reconnect_timeout_ms) * 1ms, &cntx_);
      if (ec) {
```
What am I missing here ?

Reply:
because when starting replication (Replica::Start) we call ConnectAndAuth and than state_mask_.store(R_ENABLED | R_TCP_CONNECTED);
Therefor when entering the while loop at the first time we will not execute this if condition